main/src/library/Foldable.flix

```diff
@@ -0,0 +1,165 @@
+        Foldable.foldLeft((_, _) -> false, true, t)
+
+    ///
+    /// Returns `true` if and only if the element `x` is in `t`.
+    ///
+    pub def memberOf[t :# Type -> Type : Foldable, a: Eq](x: a, t: t[a]): Bool =
+        Foldable.foldLeft((acc, y) -> acc or x == y, false, t)
+
+    ///
+    /// Optionally finds the smallest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def minimum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> Order.min(z, v), x, acc)), None, t)
+
+    ///
+    /// Optionally finds the smallest element of `t` according to the given comparator `cmp`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def minimumBy[t :# Type -> Type : Foldable, a](cmp: (a, a) -> Comparison, t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> Order.minBy(cmp, z, v), x, acc)), None, t)
+
+    ///
+    /// Optionally finds the largest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def maximum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> Order.max(z, v), x, acc)), None, t)
+
+    ///
+    /// Optionally finds the largest element of `t` according to the given comparator `cmp`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def maximumBy[t :# Type -> Type : Foldable, a](cmp: (a, a) -> Comparison, t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> Order.maxBy(cmp, z, v), x, acc)), None, t)
+
+    ///
+    /// Optionally applies `f` to all elements in `t` going from left to right until a single value is obtained.
+    ///
+    /// That is, the result is of the form: `Some(f(...f(f(x1, x2), x3)..., xn))`
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def reduceLeft[t :# Type -> Type : Foldable, a, e](f: (a, a) -> a & e, t: t[a]): Option[a] & e =
+        foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> f(v, z), x, acc)), None, t)
+
+    ///
+    /// Optionally applies `f` to all elements in `t` going from right to left until a single value is obtained.
+    ///
+    /// That is, the result is of the form: `Some(f(x1, ...f(xn-2, f(xn-1, xn))...))`
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def reduceRight[t :# Type -> Type : Foldable, a, e](f: (a, a) -> a & e, t: t[a]): Option[a] & e =
+        foldRight((x, acc) -> Some(Option.foldLeft((v, z) -> f(v, z), x, acc)), None, t)
+
+    ///
+    /// Returns `t` as an immutable list.
+    ///
+    /// That is, the result is of the form: `t1 :: t2 :: ... :: tn :: Nil`.
+    ///
+    pub def toList[t :# Type -> Type : Foldable, a](t: t[a]): List[a] =
+        List.reverse(foldLeft((acc, x) -> x :: acc, Nil, t))
```

Comment:
Or just straight-up foldRight with cons :)

Reply:
B-but stack safety...

Reply:
Hmm. But that will come later. Its a separate problem.

Reply:
Roger ;)